contracts/MarketOracle.sol

```diff
@@ -14,104 +14,128 @@ interface IOracle {
 /**
  * @title Market Oracle
  *
- * @dev Provides the aggregated exchange rate data onchain using data from a whitelisted
- *      set of market source contracts.
- *      Exchange rate is the TOKEN:TARGET rate.
+ * @dev Provides the aggregated rate data onchain using data from a whitelisted
+ *      set of providers.
  */
 contract MarketOracle is Ownable, IOracle {
     using SafeMath for uint256;
 
-    // Whitelist of sources
-    MarketSource[] public whitelist;
+    struct Report {
+        address provider;
+        uint256 timestamp;
+        uint256 payload;
+    }
 
-    event LogSourceAdded(MarketSource source);
-    event LogSourceRemoved(MarketSource source);
-    event LogSourceExpired(MarketSource source);
+    mapping (address => uint256) internal whitelistMap;
+    Report[] public whitelistReports;
+
+    event LogSourceAdded(address source);
+    event LogSourceRemoved(address source);
+    event LogSourceExpired(address source);
+
+
+    // The number of seconds after which the report must be deemed expired.
+    uint256 public reportExpirationTimeSec = 6*60*60;
+
+    uint256 public minimumProviders = 1;
+
+    function setReportExpirationTimeSec(uint256 reportExpirationTimeSec_)
+    external
+    onlyOwner
+    {
+        require(reportExpirationTimeSec_ > 0);
+        reportExpirationTimeSec = reportExpirationTimeSec_;
+    }
+
+    function setMinimumProviders(uint256 minimumProviders_)
+    external
+    onlyOwner
+    {
+        require(minimumProviders_ > 0);
+        minimumProviders = minimumProviders_;
+    }
+
+    function lookupProviderIndex(address provider) internal view returns (uint256) {
+        uint256 position = whitelistMap[provider];
+        require(position != 0);
+        return position.sub(1);
+    }
+
+    function setProviderIndex(address provider, uint256 index) internal {
+        whitelistMap[provider] = index.add(1);
+    }
+
+    function pushReport(uint256 payload) external
+    {
+        uint256 position = lookupProviderIndex(msg.sender);
+        Report storage report = whitelistReports[position];
```

Comment:
I bet it's more efficient to write it this way (but I'm not sure):
`Report storage report = Report({timestamp: block.timestamp, payload: payload});`

I can't remember the gas accounting on storage writes, like if there's a fixed cost per operation in addition to the cost per byte.

They use this style here: https://solidity.readthedocs.io/en/v0.4.24/types.html?highlight=struct

Reply:
Assigning this ways look cleaner but the problem is that it will need to write an extra SSTORE to not overwrite the address.
whitelistReports[position] = Report({timestamp: block.timestamp, payload: payload, provider: whitelistReports[position].address});
AFAIK there is no fixed cost per operation and it word is accounted for separately.